Product Collection: Enable Context-Aware Previews by Adding `usesReference` to `registerProductCollection`

imanish003 opened the pull request:
### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #47964 

In this PR, we are adding a new argument called `usesReference` to `registerProductCollection`. This argument will allow collections to specify the required context. When the required context isn't available on Editor side but will be available in Frontend, we will show a preview label.

Here is how `registerProductCollection` will look like:

```tsx
__experimentalRegisterProductCollection({
  ...,
  usesReference: ['product'],
)
```

`usesReference` can have following values:
- product
- archive
- cart
- order

`usesReference` will be used to implement multiple features in the future. Therefore, it's important to define what's part of this PR and what's not.

**What's part of this PR:**
- Adding `usesReference` argument to `registerProductCollection`.
- When reference isn't available on Editor side, but will be available on Frontend, we will show a preview label.
  - When `usesReference` is set to `product`, we will show a message "Actual products will vary depending on the product being viewed."
  - When `usesReference` is set to `archive`, we will show a message "Actual products will vary depending on the page being viewed."

**What's not part of this PR:**
- Linking a product with collection (#44877): When usesReference is set to product and product context isn't available, we will show a prompt to select a product. This way, product context will be manually selected. 
- Handling deleted product (#44878): When product context is selected manually and product is deleted, we will need to handle this case. Communicating with merchant that product is deleted and new product needs to be selected for collection to work properly.
- Error state when collection is added to unsupported context (#44879): When it's not possible to manually select a context, we will show an error state. We will ask merchant to change the collection using error state.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

**Pre-requisite:** Install this plugin - 
[register-product-collection-tester.zip](https://github.com/user-attachments/files/16362323/register-product-collection-tester.zip)


**Case 1: `usesReference: ['product']` on Single Product template**

1. Go to "Single Product" template in Editor
2. Add "My Custom Collection - Product Context" block
3. Select the block
4. Verify that you can see preview label with message "Actual products will vary depending on the product being viewed." See screenshot below.

![image](https://github.com/user-attachments/assets/e296524f-ee08-4018-bed2-8a1b3adfe94b)

**Case 2: `usesReference: ['product']` at other places**

1. For other places i.e.:
  - Specific single product template
  - Other templates like Product Catalog, Product by Category etc.
  - Post or Page
2. Add "My Custom Collection - Product Context" block
3. Smoke test to verify that preview label isn't shown.
  - **Note:** These cases don't require a preview label. As part of #44877, we will show a prompt to select a product.

**Case 3: `usesReference: ['cart']`**

1. Go to "Page: Cart" template in Editor
2. Add "My Custom Collection - Cart Context" block
3. Select the block
4. Verify that you can see preview label with message "Actual products will vary depending on the page being viewed." See screenshot below.
5. Follow 1-4 steps for:
  - "Page: Checkout" template
  - Add "My Custom Collection - Cart Context" block inside "Cart" & "Mini cart" blocks
    - **Note:** I didn't mention Checkout block because i couldn't add "My Custom Collection - Cart Context" block inside Checkout block. But if you know how you can do it, please feel free to test with Checkout block as well.
6. For all other places, smoke test to verify that preview label isn't shown.

![image](https://github.com/user-attachments/assets/d4fa7a54-b44f-4204-af30-6d881cc6cf15)

**Case 4: `usesReference: ['order']`**

1. Go to "Order Confirmation" template in Editor
2. Add "My Custom Collection - Order Context" block
3. Select the block
4. Verify that you can see preview label with message "Actual products will vary depending on the page being viewed." 
5. For all other places, smoke test to verify that preview label isn't shown.

**Case 5: `usesReference: ['archive']`**

1. Go to any of the archive template in Editor i.e. Product by Category, Product by Tag, Products by Attribute etc.
2. Add "My Custom Collection - Archive Context" block
3. Select the block
4. Verify that you can see preview label with message "Actual products will vary depending on the page being viewed."
5. For all other places, smoke test to verify that preview label isn't shown.

**Case 6: `usesReference: ['product', 'order']` - Multiple values**

1. Go to "Single Product" template in Editor
2. Add "My Custom Collection - Multiple Contexts" block
3. Select the block
4. Verify that you can see preview label with message "Actual products will vary depending on the product being viewed."
5. Follow 1-4 steps for "Order Confirmation" template too.
6. For all other places, smoke test to verify that preview label isn't shown.

**Case 7: Advance testing (to be tested by developers only)**

Feel free to take a look at plugin code in `collections.js` file. 
- You can find all the collections of above cases in this file.
- Feel free to make some changes in collections and test how it works.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [x] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Product Collection: Enable Context-Aware Previews by Adding `usesReference` to `registerProductCollection`
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

imanish003 committed:
Allow registering custom collections with various context references

Implemented the registration of product collections with different context references using the `usesReference` property. These custom collections are now registered with specific contexts such as product, cart, order, and archive.

This enhancement allows for more tailored previews and functionality based on the specific context in which the product collection is used.

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -25,6 +25,7 @@ export interface ProductCollectionConfig extends BlockVariation {
 		setPreviewState?: SetPreviewState;
 		initialPreviewState?: PreviewState;
 	};
+	usesReference?: string;
```
## Question: Should we allow multiple values i.e. array of strings for `usesReference`?

**Point 1:**
For example, if `usesReference` is set to `['product', 'order']`, what should be the message in this case? I believe depending on the context, we will show a different message. If collection is inside Single Product template, we will show a message of product context. If collection is inside Order Confirmation template, we will show a message of order context.

**Point 2:**
Just to confirm, when we allow multiple values, it will be an OR condition. Right? This means that "one of" references is required for the collection to work properly. Right? 

**Point 3:**
We will also implement a prompt to select a product when the product context isn't available. But when there are multiple values, when should we show this prompt? I believe we should show this prompt when none of the contexts is available. Right?

Also, In the future, if we plan to implement a picker for archive context, how will it work when multiple values are allowed? Will product picker have more precedence over archive picker? Maybe I am overthinking it. We will know when the time comes? 😅

**Point 4:**

Can you think of collections that may require multiple values for `usesReference`? If not, maybe we can allow a single value for `usesReference` for now. Later, we can add support for multiple values if needed.

xristos3490 replied:
> Can you think of collections that may require multiple values for usesReference?

I'm not completely certain, as this relates to the product side, but I think some key scenarios that necessitate multiple values here are the _Up-Sells_ and _Cross-Sells_ under Linked Products on the product editor. 

![Screenshot 2024-07-23 at 4 13 55 PM](https://github.com/user-attachments/assets/6c5efcb3-70e9-4dfa-be92-1cae3c88efd6)

These are "product collections" provided by core, and I understand that the intention is to use them in _both_ the **product** and **cart** contexts. This involves (a) displaying the products as expected in the product context and (b) using a combined list when in the cart, based on the cart contents.

cc @jarekmorawski  and @manospsyx to double-check

kmanijak replied:
> Point 2:
> Just to confirm, when we allow multiple values, it will be an OR condition. Right? This means that "one of" references is required for the collection to work properly. Right?

Exactly.

> Point 3: I believe we should show this prompt when none of the contexts is available. Right?

That's how I see it working.

> Point 4: Can you think of collections that may require multiple values for usesReference?

Multi-choice is a must from my POV. Collections that work with single product references, can work with multiple product references (like "cart", "order"). And that's visible in visibility matrix:

<img width="1214" alt="image" src="https://github.com/user-attachments/assets/c67aa4ec-5a78-4010-a5d5-0941823e01ec">

xristos3490 replied:
Thanks @kmanijak! That list was exactly what I was looking for! 

Re **Point 2**

Yes! That's an OR.

Re **Point 3**

Here are some additional thoughts on this:
- We should only ask for the product when the collection uses a product reference and it's not globally provided, such as in single-product templates and the single-product block.
- It's important to note that the main checks relate to whether (a) a reference is used, (b) if this reference is available, and (c) if we can manually add it. If all three are false, then an "internal" preview system must be applied. This means that any function or initial state passed into that collection (for preview) from 3PDs **should _not_ run or affect the editor until and unless the reference is provided completely.**

xristos3490 replied:
In other words, it is crucial to provide the necessary data when the `usesReference` prop requires it. The `rest_product_query` filter should only be executed, allowing 3PDs to utilize it, when the reference is completed.

imanish003 replied:
> Multi-choice is a must from my POV. Collections that work with single product references, can work with multiple product references (like "cart", "order"). And that's visible in visibility matrix:

@kmanijak Aha! Thanks for sharing the screenshot of the visibility matrix. Specifically, I like that it has specified "Visible & can use context" legend 🙌🏻

> We should only ask for the product when the collection uses a product reference and it's not globally provided, such as in single-product templates and the single-product block.

> should not run or affect the editor until and unless the reference is provided completely.

@xristos3490 💯 I agree with both of these points.

---

I will go ahead and make the changes to allow multiple values for `usesReference`. Thank you so much for the feedback. 🙇‍♂️

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -25,6 +25,7 @@ export interface ProductCollectionConfig extends BlockVariation {
 		setPreviewState?: SetPreviewState;
 		initialPreviewState?: PreviewState;
 	};
+	usesReference?: string;
```
As discussed already, that should be `string[]`

kmanijak reviewed (changes requested):
I left some feedback regarding your questions.

I think the main change should we to move from a single string value of `usesReference` to an array of strings and that change should be propagated through the whole PR. 🙌 

Let me know if you disagree.

imanish003 committed:
Support multiple references for usesReference

imanish003 committed:
docs: update documentation to include `usesReference`

Enhanced the `register-product-collection` documentation to include details about the new `usesReference` property. This property allows collections to specify required references, enhancing their contextual relevance and preview capabilities.

- Added `usesReference` description to the collection fields section.
- Included examples demonstrating the usage of `usesReference` with single and multiple references.
- Clarified the behavior when required references are unavailable on the editor side but available on the frontend.

This documentation update provides third-party developers with the necessary information to utilize the `usesReference` property effectively in their custom product collections.

imanish003 committed:
test: add e2e tests for `usesReference`

Added end-to-end tests to verify the functionality of the `usesReference` property in product collections. These tests ensure that the appropriate preview labels are displayed.

- Created multiple custom product collections with different `usesReference` values: `product`, `cart`, `order`, `archive`, and combinations of these.
- Verified that collections show the correct preview label when the required reference is present.
- Ensured that the preview label is not shown in irrelevant contexts, such as posts or product catalog templates.

This addition enhances test coverage and ensures the robustness of the `usesReference` feature.

imanish003 committed:
Fix merge conflicts

imanish003 commented:
Hey @kmanijak @xristos3490 👋 

I have updated the PR with:
- E2E tests
- Documentation for `usesReference` argument in `registerProductCollection`
- Supporting multiple values for `usesReference`

Can you please review the PR again? 🙇‍♂️

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx:
```diff
@@ -154,18 +154,71 @@ export const addProductCollectionToQueryPaginationParentOrAncestor = () => {
+			( location.type === LocationType.Archive &&
+				location.sourceData?.termId ) ||
+			( location.type === LocationType.Product &&
+				location.sourceData?.productId )
```
I don't know if it's possible for `termId` or `productId` to be `0` but if so, we should check if it has value

```suggestion
			( location.type === LocationType.Archive &&
				location.sourceData?.termId != null ) ||
			( location.type === LocationType.Product &&
				location.sourceData?.productId != null )
```

kmanijak replied:
I see you left a thumbup but I don't see it changed 😅 if you agree, please add this check!

imanish003 replied:
I have added the check in [this commit](https://github.com/woocommerce/woocommerce/pull/49796/commits/ae96a19fbbd4c43a9279c8cca21a93445f80e347). Isn't that what you were looking for? Maybe I am missing something. 😅

For example, When `termId` is `0`, the expression `(location.sourceData?.termId ?? null) !== null` evaluates to `true` because `0` is not equal to `null`.

kmanijak replied:
I missed "Outdated" label above the code snippet... Need to focus more 🙈 

<img width="307" alt="image" src="https://github.com/user-attachments/assets/a38a1af2-61a9-4ccf-856d-c3209d4913f7">

All good!

kmanijak commented on plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts:
```diff
@@ -1721,3 +1722,103 @@ test.describe( 'Testing registerProductCollection', () => {
+				'woocommerce/woocommerce//single-product',
+				'woocommerce/woocommerce//order-confirmation',
+			],
+		},
```
I don't understand how those custom collections test `usesReference` if they don't provide `usesReference` 😅 what am I missing?

imanish003 replied:
As you can see [here](https://github.com/woocommerce/woocommerce/blob/ae96a19fbbd4c43a9279c8cca21a93445f80e347/plugins/woocommerce-blocks/tests/e2e/plugins/register-product-collection-tester/index.js#L69-L110), each collection is providing `usesReference`. For example, "My Custom Collection - Product Context" is providing `usesReference: [ 'product' ]`. Were you asking about something else? 🤔

kmanijak replied:
> As you can see here, each collection is providing `usesReference`

My bad 🙈 I thought `MY_REGISTERED_COLLECTIONS` contains collection definitions. Apologies, didn't analyse that deep enough 😅

kmanijak reviewed (changes requested):
Thanks a lot for this piece of work!

When testing Product Context collection I found out it displays Preview label:
- I'm in Product: Shirt template
- Added "My Custom Collection - Product Context"
- Preview label is there

<img width="1240" alt="image" src="https://github.com/user-attachments/assets/ad70124c-9c5e-4c06-9bfc-d5e234a86e2d">

I think it may be related to the `productId` in location resolving in second iteration (initially it's `null` and `getEntityRecords` resolves later to specific value which updates the state). The same happens in specific category template. "My Custom Collection - Archive Context" displays Preview in a specific category template.

But actually similar happens in Single Product block:
- you can see "My Custom Collection - Product Context" is in Single Product block
- location is recognized correctly (visible in request data)
- Preview label is there

![Screen Shot 2024-07-26 at 10 05 20 AM](https://github.com/user-attachments/assets/663201e0-e785-4176-999d-9aa30d375dbd)

Others scenarios seem to be working fine :)

imanish003 committed:
Fixed issue with preview label showing in Specific Product and Archive templates.

imanish003 committed:
Handle cases when termId or productId is 0

This also improves the readability of the code by using better variable names.

imanish003 commented:
Hey @kmanijak! Thanks for thorough review 🙌🏻

I have:
- ✅ Fixed issue with preview label showing in Specific Product and Archive templates.
- ✅ Handled case when termId or productId could be 0.

---

> But actually similar happens in Single Product block:

Regarding this, looks like there is a bug because of which location isn't recognized correctly in Single Product block. FWIW I verified this bug in trunk as well. Are you able to reproduce this issue in trunk as well? 

I have created a [draft PR](https://github.com/woocommerce/woocommerce/pull/50051) to fix this issue. If you are able to reproduce this issue in trunk as well, I would recommend to test this PR and see if it fixes the issue. If it does, I will mark this draft PR as ready for review. 🙂

kmanijak committed:
Make PC use postId context so location can be recognised correctly (#50152)

kmanijak reviewed (approved):
> Regarding this, looks like there is a bug because of which location isn't recognized correctly in Single Product block. FWIW I verified this bug in trunk as well. Are you able to reproduce this issue in trunk as well?

> I have created a https://github.com/woocommerce/woocommerce/pull/50051 to fix this issue. If you are able to reproduce this issue in trunk as well, I would recommend to test this PR and see if it fixes the issue. If it does, I will mark this draft PR as ready for review. 🙂

As discussed elsewhere, the issue was a missing `postId` context in PC block which was added in https://github.com/woocommerce/woocommerce/pull/50152 and it fixed the problem of incorrect Preview label in Single Product block.

Otherwise, it all works fine! I'm approving.

I'm not convinced about one use case though. If I add a collection that requires some context to a different context (e.g. collection `usesReference: ['archive']` and is added to Single Product template) we will be displaying some information it should not work there. But until we do, I'm wondering if displaying a "Preview" label would be better than currently simply displaying products... on the other hand it's not gonna be in use straight away so I'm probably overthinking it 😅

imanish003 commented:
> I'm not convinced about one use case though. If I add a collection that requires some context to a different context (e.g. collection usesReference: ['archive'] and is added to Single Product template) we will be displaying some information it should not work there. But until we do, I'm wondering if displaying a "Preview" label would be better than currently simply displaying products... on the other hand it's not gonna be in use straight away so I'm probably overthinking it 😅

@kmanijak My understanding is that we will be tackling this in issue [#44879](https://github.com/woocommerce/woocommerce/issues/44879) in which we will show an error state when collection is added to unsupported context. Therefore, I don't think we should be handling this case in this PR. What do you think? 🤔

xristos3490 commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/block.json:
```diff
@@ -51,7 +51,7 @@
 		"__privateProductCollectionPreviewState": "__privatePreviewState"
 	},
-	"usesContext": [ "templateSlug" ],
+	"usesContext": [ "templateSlug", "postId" ],
```
Why is this context needed here? 🤔

xristos3490 replied:
Oh, never mind. I thought this was on the product-template block 😁

xristos3490 reviewed (commented):
I agree with @kmanijak that the work at https://github.com/woocommerce/woocommerce/issues/44879 is imprortant here in order for this feature to be completed. 

Another note (that we already discussed) is that this doesn't work for custom taxonomies. Do we have GH issue to track this work?

Overall, its looking good, and works as described.  Left a comment regarding the `postId` context before approving.

imanish003 commented:
> I agree with @kmanijak that the work at #44879 is imprortant here in order for this feature to be completed.

Agreed! Let's tackle #44879 in next cycle 🙂

> Another note (that we already discussed) is that this doesn't work for custom taxonomies. Do we have GH issue to track this work?

@xristos3490 Good point! I believe for this we will need to update logic in [useGetLocation](https://github.com/woocommerce/woocommerce/blob/9bda9cfb4144c411fa80d57bee034017944636ee/plugins/woocommerce-blocks/assets/js/blocks/product-template/utils.tsx#L83) to also recognize custom taxonomies. On Frontend side, I believe it will require some changes to [parse_frontend_location_context](https://github.com/woocommerce/woocommerce/blob/9bda9cfb4144c411fa80d57bee034017944636ee/plugins/woocommerce/src/Blocks/Utils/ProductCollectionUtils.php#L97) function. Right? 

IMO We can create a new issue around it but we should work on it once we have a clear understanding of what needs to be done. For now, we are still figuring out last pieces of context collections. What are your thoughts on this? 🙂

CC: @kmanijak

imanish003 committed:
Merge branch 'trunk' into add/47964-product-collection-collection-should-be-able-to-specify-requiredcontext-or-requiredlocation-and-trigger-preview-mode-in-editor-when-theres-no-required-context

imanish003 committed:
Fix TS errors

xristos3490 commented:
> On Frontend side, I believe it will require some changes to [parse_frontend_location_context](https://github.com/woocommerce/woocommerce/blob/9bda9cfb4144c411fa80d57bee034017944636ee/plugins/woocommerce/src/Blocks/Utils/ProductCollectionUtils.php#L97) function. Right?

@imanish003 No changes are needed on the PHP side since it already [uses](https://github.com/WordPress/wordpress-develop/blob/74e03e3cbef2f2565028f446c76acb2dabf749bd/src/wp-includes/meta.php#L336C15-L336C40) the `is_product_taxonomy()` helper that includes custom ones 😁

imanish003 commented:
Hey @xristos3490! I have created issue #50253. Please feel free to add more details if I missed something 🙇  CC: @kmanijak
